File: Introduction to Robotics Course/Chapter14/Lesson4/Chapter14_Lesson4.py
def update_trust(successes, lam=0.2, T0=0.5):
    """
    successes: list of 0/1 values (perceived failures/successes)
    lam: trust update gain (0 < lam <= 1)
    T0: initial trust level
    """
    T = T0
    history = [T]
    for s in successes:
        T = (1.0 - lam) * T + lam * float(s)
        history.append(T)
    return history

if __name__ == "__main__":
    # Example: 1=success, 0=failure over 10 interactions
    seq = [1, 1, 0, 1, 1, 1, 0, 1, 1, 1]
    trust_traj = update_trust(seq, lam=0.3, T0=0.2)
    for k, T in enumerate(trust_traj):
        print(f"Step {k}: T = {T:.3f}")
      

File: Tutorials/Codes/Introduction_to_Robotics_Course/Chapter14/Lesson4/Chapter14_Lesson4.java
public class TrustFilter {
    public static double[] updateTrust(int[] successes, double lambda, double T0) {
        double T = T0;
        double[] history = new double[successes.length + 1];
        history[0] = T;
        for (int k = 0; k < successes.length; ++k) {
            T = (1.0 - lambda) * T + lambda * (double) successes[k];
            history[k + 1] = T;
        }
        return history;
    }

    public static void main(String[] args) {
        int[] seq = {1, 1, 0, 1, 1, 1, 0, 1, 1, 1};
        double lambda = 0.3;
        double T0 = 0.2;
        double[] trustTraj = updateTrust(seq, lambda, T0);

        for (int k = 0; k < trustTraj.length; ++k) {
            System.out.printf("Step %d: T = %.3f%n", k, trustTraj[k]);
        }
    }
}
      
